lib/lists.nix

```diff
@@ -403,6 +403,45 @@ rec {
   */
   concatMap = builtins.concatMap;
 
+  /**
+    Map, removing null values.
+
+    # Type
+
+    ```
+    filterMap :: (a -> [b]) -> [a] -> [b]
+    ```
+
+    # Examples
+    :::{.example}
+    ## `lib.lists.filterMap` usage example
+
+    ```nix
+    let f = x:
+      if "udp" == x.proto
+      then x.port
+      else null;
+    in filterMap f [
+      {
+        dns.proto  = "udp";
+        dns.port  = 53;
+      }
+      {
+        http.proto = "tcp";
+        http.port = 80;
+      }
+    ]
+    => [ 53 ]
+    ```
+
+    :::
+  */
+  filterMap = f:
+    let g = x: acc:
+      let y = f x;
+      in if isNull y then acc else [y] ++ acc;
```

Comment:
```suggestion
      in if y == null then acc else [y] ++ acc;
```

Same number of AST nodes, but simpler operations.

```
$ bench -- 'nix eval --impure --expr "let lib = import ./lib; inherit (lib) foldl'\'' range; inherit (builtins) isNull; in foldl'\'' (a: b: a || isNull b) false (range 0 10000000)"';                                   bench -- 'nix eval --impure --expr "let lib = import ./lib; inherit (lib) foldl'\'' range; inherit (builtins) isNull; in foldl'\'' (a: b: a || b == null) false (range 0 10000000)"'
benchmarking nix eval --impure --expr "let lib = import ./lib; inherit (lib) foldl' range; inherit (builtins) isNull; in foldl' (a: b: a || isNull b) false (range 0 10000000)"
time                 2.014 s    (NaN s .. 2.022 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 2.010 s    (2.008 s .. 2.012 s)
std dev              2.497 ms   (1.883 ms .. 2.958 ms)
variance introduced by outliers: 19% (moderately inflated)

benchmarking nix eval --impure --expr "let lib = import ./lib; inherit (lib) foldl' range; inherit (builtins) isNull; in foldl' (a: b: a || b == null) false (range 0 10000000)"
time                 1.938 s    (1.915 s .. 1.952 s)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 1.941 s    (1.938 s .. 1.944 s)
std dev              3.180 ms   (1.610 ms .. 4.395 ms)
variance introduced by outliers: 19% (moderately inflated)

```

Reply:
Also strictness is not an issue. The type tag is compared first, and there won't be any need to evaluate any nested values, because those are only evaluated if the type tag matches, and if it matches, it is `null`, which has no nested values.